test/js-api/limits.js

```diff
@@ -0,0 +1,340 @@
+            builder.addDataSegment(0, data);
+        }
+    });
+
+testLimit("initial declared memory pages", 1, kJSEmbeddingMaxMemoryPages,
+          (builder, count) => {
+            builder.addMemory(count, undefined, false, false);
+          });
+
+testLimit("maximum declared memory pages", 1, kJSEmbeddingMaxMemoryPages,
+          (builder, count) => {
+            builder.addMemory(1, count, false, false);
+          });
+
+testLimit("initial imported memory pages", 1, kJSEmbeddingMaxMemoryPages,
+          (builder, count) => {
+            builder.addImportedMemory("mod", "mem", count, undefined);
+          });
+
+testLimit("maximum imported memory pages", 1, kJSEmbeddingMaxMemoryPages,
+          (builder, count) => {
+            builder.addImportedMemory("mod", "mem", 1, count);
+          });
+
+// TODO(titzer): ugh, that's hard to test.
+DISABLED.testLimit("module size", 1, kJSEmbeddingMaxModuleSize,
+                   (builder, count) => {
+                   });
+
+testLimit("function size", 2, kJSEmbeddingMaxFunctionSize, (builder, count) => {
+        let type = builder.addType(kSig_v_v);
+        let nops = count-2;
+        let array = new Array(nops);
+        for (let i = 0; i < nops; i++) array[i] = kExprNop;
+        builder.addFunction(undefined, type).addBody(array);
+    });
+
+testLimit("function locals", 1, kJSEmbeddingMaxFunctionLocals, (builder, count) => {
+        let type = builder.addType(kSig_v_v);
+        builder.addFunction(undefined, type)
+          .addLocals({i32_count: count})
+          .addBody([]);
+    });
+
+testLimit("function params", 1, kJSEmbeddingMaxFunctionParams, (builder, count) => {
+        let array = new Array(count);
+        for (let i = 0; i < count; i++) array[i] = kWasmI32;
+        let type = builder.addType({params: array, results: []});
+    });
+
+testLimit("function params+locals", 1, kJSEmbeddingMaxFunctionLocals - 2, (builder, count) => {
+        let type = builder.addType(kSig_i_ii);
+        builder.addFunction(undefined, type)
+          .addLocals({i32_count: count})
+          .addBody([kExprUnreachable]);
+    });
+
+testLimit("function returns", 0, kJSEmbeddingMaxFunctionReturns, (builder, count) => {
+        let array = new Array(count);
+        for (let i = 0; i < count; i++) array[i] = kWasmI32;
+        let type = builder.addType({params: [], results: array});
+    });
+
+testLimit("initial table size", 1, kJSEmbeddingMaxTableSize, (builder, count) => {
+        builder.setFunctionTableBounds(count, undefined);
+    });
+
+testLimit("maximum table size", 1, kJSEmbeddingMaxTableSize, (builder, count) => {
+        builder.setFunctionTableBounds(1, count);
+    });
+
+testLimit("table entries", 1, kJSEmbeddingMaxTableEntries, (builder, count) => {
+        builder.setFunctionTableBounds(1, 1);
+        let array = [];
+        for (let i = 0; i < count; i++) {
+            builder.addFunctionTableInit(0, false, array, false);
```

Comment:
This terminology confused me at first; "FunctionTableInit" here is what the spec calls an "Element Segment". I've opened a PR to rename these: https://github.com/WebAssembly/spec/pull/874

Reply:
ack